tonic/src/transport/service/tls.rs

```diff
@@ -56,41 +56,67 @@ enum Connector {
 
 impl TlsConnector {
     #[cfg(feature = "openssl")]
-    pub(crate) fn new_with_openssl(
-        cert: Certificate,
+    pub(crate) fn new_with_openssl_cert(
+        cert: Option<Certificate>,
         domain: String,
     ) -> Result<Self, crate::Error> {
         let mut config = SslConnector::builder(SslMethod::tls())?;
-
         config.set_alpn_protos(ALPN_H2_WIRE)?;
 
-        let ca = X509::from_pem(&cert.pem[..])?;
-
-        config.cert_store_mut().add_cert(ca)?;
+        if cert.is_some() {
+            let cert = cert.unwrap();
+            let ca = X509::from_pem(&cert.pem[..])?;
+            config.cert_store_mut().add_cert(ca)?;
+        }
 
-        let config = config.build();
+        //        let config = config.build();
 
         Ok(Self {
-            inner: Connector::Openssl(config),
+            inner: Connector::Openssl(config.build()),
             domain: Arc::new(domain),
         })
     }
 
-    #[cfg(feature = "rustls")]
-    pub(crate) fn new_with_rustls(cert: Certificate, domain: String) -> Result<Self, crate::Error> {
-        let mut buf = std::io::Cursor::new(&cert.pem[..]);
+    #[cfg(feature = "openssl")]
+    pub(crate) fn new_with_openssl_raw(
+        ssl_connector: openssl1::ssl::SslConnector,
+    ) -> Result<Self, crate::Error> {
+        Ok(Self {
+            inner: Connector::Openssl(ssl_connector),
+            domain: Arc::new("".into()),
+        })
+    }
 
+    #[cfg(feature = "rustls")]
+    pub(crate) fn new_with_rustls_cert(
+        cert: Option<Certificate>,
+        domain: String,
+    ) -> Result<Self, crate::Error> {
         let mut config = ClientConfig::new();
-
-        config.root_store.add_pem_file(&mut buf).unwrap();
         config.set_protocols(&[Vec::from(&ALPN_H2[..])]);
 
+        if cert.is_some() {
+            let cert = cert.unwrap();
+            let mut buf = std::io::Cursor::new(&cert.pem[..]);
+            config.root_store.add_pem_file(&mut buf).unwrap();
+        }
+
         Ok(Self {
             inner: Connector::Rustls(Arc::new(config)),
             domain: Arc::new(domain),
         })
     }
 
+    #[cfg(feature = "rustls")]
+    pub(crate) fn new_with_rustls_raw(
+        config: tokio_rustls::rustls::ClientConfig,
+    ) -> Result<Self, crate::Error> {
+        Ok(Self {
+            inner: Connector::Rustls(Arc::new(config)),
+            domain: Arc::new("".into()),
```

Comment:
do you think an option might be better here than an empty string?

Reply:
Actually it turns out this wouldn't have worked as implemented anyway: the domain name is required even with a raw config. That has simplified this area.